### PR TITLE
Added the Web Consumer using the HostedService

### PR DESCRIPTION
**Pull request by dhavalsp**
I have create WebConsumer example using the IHostedService to develop the Web Cosumer.

This is my first time contribution experience. Please pardon me any mistake. Let me know to make any changes.

### CODE REVIEW

**Comment by dhavalsp** [comment]
Web Consumer using IHostedService

**Comment by dhavalsp** [comment]
[clabot:check]

**Comment by AndyPook** [comment]
Nice, but...

Prefer `Microsoft.Extensions.Hosting.BackgroundService` from Microsoft.Extensions.Hosting.Abstractions (probably already included)

File: examples/WebConsumer/WebConsumer.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        private readonly static string _BOOTSRAP_SERVER = "localhost:9092"; // BROKER address
+        private readonly static string _TOPIC = "YourTopicName"; // Topic name
+        private readonly static string _GroupId = "GroupName";
+
+        static ConsumerConfig config;
+        static IConsumer<string, string> consumer;
+
+        public WebConsumer()
+        {
+            config = new ConsumerConfig
```
Comment (mhowlett) [inline examples/WebConsumer/WebConsumer.cs]: should be via IConfiguration

File: examples/WebConsumer/WebConsumer.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            consumer = new ConsumerBuilder<string, string>(config).Build();
+        }
+
+        public async Task Start()
+        {
+            try
+            {
+                consumer.Subscribe(_TOPIC);
+                CancellationTokenSource cts = new CancellationTokenSource();
+                Console.CancelKeyPress += (_, e) =>
```
Comment (mhowlett) [inline examples/WebConsumer/WebConsumer.cs]: seems out of place

File: examples/WebConsumer/WebConsumer.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                {
+                    e.Cancel = true; // prevent the process from terminating.
+                    cts.Cancel();
+                };
+
+                var consumeResult = consumer.Consume(cts.Token);
+                Debug.WriteLine($"Reached end of topic {consumeResult.Topic}, partition {consumeResult.Partition}, offset {consumeResult.Offset}.");
```
Comment (mhowlett) [inline examples/WebConsumer/WebConsumer.cs]: not necessarily at end of topic.

**Comment by mhowlett** [comment]
thanks @dhavalsp - one of the activities i have on the go at the moment is making some good abstractions to allow `Confluent.Kafka` to be used in web applications. stay tuned!

**Comment by bbjoern** [comment]
@dhavalsp great thanks for your input! sad to say that i'm still stuck with the same problem I had with my version as BackgroundService. It seems like "consumer.Consume(..)" breaks the whole webapp.

Here was my Version with BackgroundService (registered in Startup via services.AddHostedService<ConsumerService >(); )

`
public class ConsumerService : BackgroundService {

        private readonly ConsumerConfig _consumerConfig;

        public UserTaskService() {
            this._mediator = mediator;
            this._hubContext = hubContext;
            this._consumerConfig = new ConsumerConfig {
                BootstrapServers = "kafka:9092",
                GroupId = "task-consumer",
                EnableAutoCommit = true,
                StatisticsIntervalMs = 5000,
                SessionTimeoutMs = 6000,
                AutoOffsetReset = AutoOffsetReset.Earliest,
                EnablePartitionEof = true
            };
        }

        protected override async Task ExecuteAsync(CancellationToken stoppingToken) {
            Console.WriteLine("OrderProcessing Service Started");

            while (!stoppingToken.IsCancellationRequested) {

                using(var consumer = new ConsumerBuilder<Ignore, string>(this._consumerConfig)
                    .SetErrorHandler((_, e) => Console.WriteLine($"Error: {e.Reason}"))
                    .SetStatisticsHandler((_, json) => Console.WriteLine($"Statistics: {json}"))
                    .SetPartitionsAssignedHandler((c, partitions) => {
                        Console.WriteLine($"Assigned partitions: [{string.Join(", ", partitions)}]");
                    })
                    .SetPartitionsRevokedHandler((c, partitions) => {
                        Console.WriteLine($"Revoking assignment: [{string.Join(", ", partitions)}]");
                    })
                    .Build()) {

                    consumer.Subscribe(new List<string>() { "oozima-tasks" });  
                    var consumeResult = consumer.Consume(stoppingToken);

                    if (consumeResult.IsPartitionEOF) {
                        Console.WriteLine(
                            $"Reached end of topic {consumeResult.Topic}, partition {consumeResult.Partition}, offset {consumeResult.Offset}.");

                        continue;
                    }

                    Console.WriteLine($"Received message at {consumeResult.TopicPartitionOffset}: {consumeResult.Value}");
                }
            }
        }
    }`

@mhowlett did you invest further in this ?

**Comment by dhavalsp** [comment]
I apologize for late reply. I was little busy and out of town. I was facing same issues but after so many try, I was successes. Honestly, I am not sure at first glance. but, I would try small demo with your code. Let me see if i would find something.
But, Mean while you can try this below hint:
I created the WebConsumerExecutor class and WebConsumer class separately. I also added the singleton in startup.cs like this.
```
   services.AddSingleton<WebConsumer>();
   services.AddSingleton<IHostedService,WebConsumerExecutor>();
```

I am trying with your code too.

**Comment by dhavalsp** [comment]
> @dhavalsp great thanks for your input! sad to say that i'm still stuck with the same problem I had with my version as BackgroundService. It seems like "consumer.Consume(..)" breaks the whole webapp.
> 
> Here was my Version with BackgroundService (registered in Startup via services.AddHostedService(); )
> 
> ```
> public class ConsumerService : BackgroundService {
>         private readonly ConsumerConfig _consumerConfig;
> 
>         public ConsumerService() {
>             this._mediator = mediator;
>             this._hubContext = hubContext;
>             this._consumerConfig = new ConsumerConfig {
>                 BootstrapServers = "kafka:9092",
>                 GroupId = "task-consumer",
>                 EnableAutoCommit = true,
>                 StatisticsIntervalMs = 5000,
>                 SessionTimeoutMs = 6000,
>                 AutoOffsetReset = AutoOffsetReset.Earliest,
>                 EnablePartitionEof = true
>             };
>         }
> 
>         protected override async Task ExecuteAsync(CancellationToken stoppingToken) {
>             Console.WriteLine("OrderProcessing Service Started");
> 
>             while (!stoppingToken.IsCancellationRequested) {
> 
>                 using(var consumer = new ConsumerBuilder<Ignore, string>(this._consumerConfig)
>                     .SetErrorHandler((_, e) => Console.WriteLine($"Error: {e.Reason}"))
>                     .SetStatisticsHandler((_, json) => Console.WriteLine($"Statistics: {json}"))
>                     .SetPartitionsAssignedHandler((c, partitions) => {
>                         Console.WriteLine($"Assigned partitions: [{string.Join(", ", partitions)}]");
>                     })
>                     .SetPartitionsRevokedHandler((c, partitions) => {
>                         Console.WriteLine($"Revoking assignment: [{string.Join(", ", partitions)}]");
>                     })
>                     .Build()) {
> 
>                     consumer.Subscribe(new List<string>() { "oozima-tasks" });  
>                     var consumeResult = consumer.Consume(stoppingToken);
> 
>                     if (consumeResult.IsPartitionEOF) {
>                         Console.WriteLine(
>                             $"Reached end of topic {consumeResult.Topic}, partition {consumeResult.Partition}, offset {consumeResult.Offset}.");
> 
>                         continue;
>                     }
> 
>                     Console.WriteLine($"Received message at {consumeResult.TopicPartitionOffset}: {consumeResult.Value}");
>                 }
>             }
>         }
>     }
> ```
> @mhowlett did you invest further in this ?

I apologize for late reply. I was little busy and out of town. I was facing same issues but after so many try, I was successes. Honestly, I am not sure at first glance. but, I would try small demo with your code. Let me see if i would find something.
But, Mean while you can try this below hint:
I created the WebConsumerExecutor class and WebConsumer class separately. I also added the singleton in startup.cs like this.
```
   services.AddSingleton<WebConsumer>();
   services.AddSingleton<IHostedService,WebConsumerExecutor>();
```

I am trying with your code too.

**Comment by bbjoern** [comment]
@dhavalsp thanks for your reply! sadly still doesn't works for me .. always stuck at consumer.Consume

**Comment by AndyPook** [comment]
I see this is Closed but thought I'd add some info as I'm using something similar in production right now...

The thing with the `Consume` overload that takes a CancellationToken is that it will not return until there is a message on the topic. So it appears to block (although you should still be seeing the EndOfPartition message 🤔 )

The actual problem with the example above is that you are creating a new consumer *inside* the loop!
Setup the consumer outside and treat the loop as a "for each message" idea.

The style we use is to inject the consumer into the service. The bootstrapping of the process (we have a set of extension methods called from Startup) creates, configures and subscribes the client. The BackgroundService is just responsible to getting a message and processing it.

```
protected override async Task ExecuteAsync(CancellationToken stoppingToken) {
    while (!stoppingToken.IsCancellationRequested) {
        var cr = consumer.Consume(100);
        if(cr != null){
            // do something to process the message, we have something like...
           // await handler.Handle(cr.Message);
        }
    }
}
```
(Abbreviated: ours has try/catch, diagnostics, metrics etc)
The 100 is the maximum ms to wait for a message. If a message is available it will return immediately.
Your version that uses the CT ought to work fine too, but internally it's just doing the same thing.
We have a whole set of framework bits that help with setting up service processes (wondering is the company will let me OSS some of it 🤔 )

**Comment by dhavalsp** [comment]
> @dhavalsp thanks for your reply! sadly still doesn't works for me .. always stuck at consumer.Consume

@AndyPook  is right. Thanks to remind me even.  I did the same thing. I use webConsumer as singleton and use as dependency Injection. Use only one object and just consume the message in the while loop.

**Comment by bbjoern** [comment]
@AndyPook Awesome that worked ! Thanks a lot !

**Pull request closed without merging**